renderdoc/driver/d3d12/d3d12_shaderdebug.cpp

```diff
@@ -420,9 +435,749 @@ ShaderDebugTrace D3D12Replay::DebugVertex(uint32_t eventId, uint32_t vertid, uin
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to close command list HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(psBlob);
+    SAFE_RELEASE(pInitialValuesBuffer);
+    SAFE_RELEASE(pInitialValuesStaging);
+    SAFE_RELEASE(pRootSignature);
+    SAFE_RELEASE(initialPso);
+    return empty;
+  }
+
+  {
+    ID3D12CommandList *l = cmdList;
+    m_pDevice->GetQueue()->ExecuteCommandLists(1, &l);
+    m_pDevice->GPUSync();
+  }
+
+  {
+    D3D12MarkerRegion initState(m_pDevice->GetQueue()->GetReal(),
+                                "Replaying event for initial states");
+
+    // Replay the event with our modified state set
+    m_pDevice->ReplayLog(0, eventId, eReplay_OnlyDraw);
+
+    cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
+
+    // Copy the initial values to the staging buffer
+    D3D12_RESOURCE_BARRIER copyToStagingBarrier;
+    ZeroMemory(&copyToStagingBarrier, sizeof(D3D12_RESOURCE_BARRIER));
+    copyToStagingBarrier.Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
+    copyToStagingBarrier.Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
+    copyToStagingBarrier.Transition.StateBefore = D3D12_RESOURCE_STATE_UNORDERED_ACCESS;
+    copyToStagingBarrier.Transition.StateAfter = D3D12_RESOURCE_STATE_COPY_SOURCE;
+    copyToStagingBarrier.Transition.pResource = pInitialValuesBuffer;
+    copyToStagingBarrier.Transition.Subresource = 0;
+    cmdList->ResourceBarrier(1, &copyToStagingBarrier);
+
+    cmdList->CopyResource(pInitialValuesStaging, pInitialValuesBuffer);
+    hr = cmdList->Close();
+    if(FAILED(hr))
+    {
+      RDCERR("Failed to close command list HRESULT: %s", ToStr(hr).c_str());
+      SAFE_RELEASE(psBlob);
+      SAFE_RELEASE(pInitialValuesBuffer);
+      SAFE_RELEASE(pInitialValuesStaging);
+      SAFE_RELEASE(pRootSignature);
+      SAFE_RELEASE(initialPso);
+      return empty;
+    }
+
+    ID3D12CommandList *l = cmdList;
+    m_pDevice->GetQueue()->ExecuteCommandLists(1, &l);
+    m_pDevice->GPUSync();
+  }
+
+  // Replaying the event has finished, and the data has been copied to the staging buffer.
+  // Free all the resources we created except for the staging buffer.
+  SAFE_RELEASE(psBlob);
+  SAFE_RELEASE(pRootSignature);
+  SAFE_RELEASE(pInitialValuesBuffer);
+  SAFE_RELEASE(initialPso);
+
+  byte *initialData = NULL;
+  pInitialValuesStaging->Map(0, NULL, (void **)&initialData);
+
+  DebugHit *buf = (DebugHit *)initialData;
+
+  D3D12MarkerRegion::Set(m_pDevice->GetQueue()->GetReal(),
+                         StringFormat::Fmt("Got %u hits", buf[0].numHits));
+  if(buf[0].numHits == 0)
+  {
+    RDCLOG("No hit for this event");
+    D3D12_RANGE emptyWriteRange = {0, 0};
+    pInitialValuesStaging->Unmap(0, &emptyWriteRange);
+    SAFE_RELEASE(pInitialValuesStaging);
+    return empty;
+  }
+
+  // Restore D3D12 state to what the event uses
+  cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
+  rs = prevState;
+  rs.ApplyState(m_pDevice, cmdList);
+  hr = cmdList->Close();
```

Comment:
What is this command list doing here? I'm not sure `ApplyState` has any side-effects if there are no draws/dispatches.

Reply:
With some testing for implementing sample/gather, I was encountering bugs where the modified state was applied instead of the original event's state. I suspect that just restoring the `D3D12RenderState` object will suffice, so I'll change to that and verify that things still work with sample/gather with testing later.

Reply:
yeh you will definitely need to restore the `m_RenderState`, since subsequent replays of the draw only (without doing a full replay from the start) will use it to know what the current partial state is. But it won't need a command list to apply it because in D3D12 the state is command buffer local, and will get discarded after it is closed.